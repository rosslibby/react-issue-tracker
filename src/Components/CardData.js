import React, { Component } from 'react';

class CardData extends Component {
    render () {

        if (this.props.link) {
            return (
                <span className="card-data">
                    <a
                        href={ this.props.url }
                        className="card-data__link"
                        target={ this.props.target }
                    >{ this.props.text }</a>
                </span>
            );
        } else {
            return (
                <span className="card-data">
                    { this.props.text }
                </span>
            );
        }
    }
};

export default CardData;