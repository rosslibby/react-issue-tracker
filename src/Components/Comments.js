import React, { Component } from 'react';

class Comments extends Component {
    createMarkup (comment) {
        return {__html: comment};
    }

    render () {

        return (
            <div>
                { this.props.comments.map((comment, index) => {
                    console.log(comment);
                    return (
                        <span
                            key={ index }
                            dangerouslySetInnerHTML={ this.createMarkup(comment.body_html) }
                        ></span>
                    );
                }) }
            </div>
        );
    }
};

export default Comments;