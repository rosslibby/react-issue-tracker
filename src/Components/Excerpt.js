import React, { Component } from 'react';

class Excerpt extends Component {

    getExcerpt (text) {
        let modifiedText = text.substr(0, 140);

        if (!(/^[a-z0-9]+$/i.test(modifiedText.substr(139, 1)) && /^[a-z0-9]+$/i.test(text.substr(140, 1))) || text.length < 140) {
            if (text.length > 140) {
                return modifiedText + '...';
            }

            return modifiedText;
        }

        return modifiedText.substr(0, modifiedText.lastIndexOf(' ')) + '...';
    }

    render () {
        return (
            <p>
                { this.getExcerpt(this.props.text) }
            </p>
        );
    }
};

export default Excerpt;