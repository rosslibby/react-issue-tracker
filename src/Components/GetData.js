import React, { Component } from 'react';
import Label from './Label';

class GetData extends Component {
    constructor (props) {
        super(props);
        this.state = { items: [] };
    }

    componentDidMount = () => {
        let xhr = new XMLHttpRequest();
        xhr.open('GET', 'https://api.github.com/repos/rails/rails/issues?per_page=25', true);
        xhr.setRequestHeader('Accept', 'application/vnd.github.text+json');
        xhr.send(null);
        xhr.onreadystatechange = () => {
            if (xhr.readyState === 4) {
                if (xhr.status === 200) {
                    this.setState({ items: JSON.parse(xhr.responseText)});
                }
            }
        };
    }

    formatLabel = (label) => {
        if (label.indexOf('https://api.github.com/repos/') !== -1) {
            let newLabel = label.replace('https://api.', 'https://');
            newLabel = newLabel.replace('.com/repos/', '.com/');

            return newLabel;
        }

        return label;
    }

    render = () => {
        return (
            <ol>
                {this.state.items.map((item) => {
                    return (
                        <li key={ item.id }>
                            <strong>{ item.id }</strong>{' | '}<strong>{ item.title }</strong>{' | '}
                            <a href={ item.html_url } target="_blank">{'[link]'}</a>
                            <p>{ item.body_text.substr(0, 140) }</p>
                            <small><em>{ item.user.login }</em></small>
                            <img src={ item.user.avatar_url + '&s=88' } alt={ item.user.login } />
                            { item.labels.map((label, index) => {
                                return (
                                    <Label
                                        key={ index }
                                        text={ label.name }
                                        url={ this.formatLabel(label.url) }
                                    />
                                );
                            }) }
                        </li>
                    );
                })}
            </ol>
        );
    }
};

export default GetData;