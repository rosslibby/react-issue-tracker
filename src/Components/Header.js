import React, { Component } from 'react';
import { Link } from 'react-router';

class Header extends Component {
    render () {
        return (
            <ul>
                <li><Link to={'/'}>Home</Link></li>
            </ul>
        );
    }
};

export default Header;