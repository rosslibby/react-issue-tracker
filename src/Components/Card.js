import React, { Component } from 'react';
import Excerpt from './Excerpt';
import Labels from './Labels';
import Title from './Title';
import User from './User';
import CardData from './CardData';

class Card extends Component {
    render () {
        const titleProps = this.props.title;
        const excerptProps = this.props.excerpt;
        const userProps = this.props.user;
        const commentProps = {
            link: false,
            text: 'Comments: ' + this.props.comments
        };
        const linkProps = {
            link: true,
            text: 'external link',
            url: this.props.url,
            target: '_blank'
        };

        return (
            <div className="card" key={ this.props.id }>
                <Title { ...titleProps } />
                <Excerpt { ...excerptProps }  />
                <div>
                    <Labels labels={ this.props.labels } />
                    <User { ...userProps } />
                    <CardData { ...commentProps } />
                    <CardData { ...linkProps } />
                </div>
            </div>
        );
    }
};

export default Card;