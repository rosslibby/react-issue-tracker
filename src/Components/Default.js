import React, { Component } from 'react';
import Header from './Header';
import List from './List';
import Pagination from './Pagination';

class Default extends Component {
    constructor (props) {
        super(props);
        this.state = { items: [], page: 0 };
    }

    componentDidMount () {
        // this.fetchPage();
    }

    fetchPage (numPages) {
        numPages = this.state.page + numPages || 1;

        let xhr = new XMLHttpRequest();
        xhr.open('GET', 'https://api.github.com/repos/rails/rails/issues?per_page=25&page=' + numPages, true);
        xhr.setRequestHeader('Accept', 'application/vnd.github.text+json');
        xhr.send(null);
        xhr.onreadystatechange = () => {
            if (xhr.readyState === 4) {
                if (xhr.status === 200) {
                    this.setState({ items: JSON.parse(xhr.responseText), page: numPages });
                    return this.state;
                }
            }
        };
    }

    render () {
        const listProps = {
            items: this.state.items
        };

        const pageProps = {
            next: () => { this.fetchPage(1) },
            prev: () => { this.fetchPage(-1) }
        };

        return (
            <div>
                <Header />
                <Pagination { ...pageProps } />
                <List { ...listProps } />
            </div>
        );
    }
};

export default Default;