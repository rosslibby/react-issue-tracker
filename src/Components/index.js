import Comments from './Comments';
import Default from './Default';
import Issue from './Issue';
import Label from './Label';
import List from './List';
import Title from './Title';
import Excerpt from './Excerpt';

module.exports = {
    Comments,
    Default,
    Excerpt,
    Issue,
    Label,
    List,
    Title
};