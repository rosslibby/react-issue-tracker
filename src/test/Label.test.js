import React from 'react';
import renderer from 'react-test-renderer';
import { Label } from '../Components/';

it('should have a link', () => {
    const label = renderer.create(<Label />);

    let tree = label.toJSON();

    expect(tree).toMatchSnapshot();
});