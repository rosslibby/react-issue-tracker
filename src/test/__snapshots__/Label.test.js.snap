exports[`test should have a link 1`] = `
<a
  className="label"
  href="#"
  target="_blank">
  link text
</a>
`;
